mac/hid.c

```diff
@@ -193,13 +196,84 @@ static struct hid_api_version api_version = {
 static	IOHIDManagerRef hid_mgr = 0x0;
 static	int is_macos_10_10_or_greater = 0;
 
+/* Global error message that is not specific to a device, e.g. for
+   hid_open(). It is thread-local like errno. */
+__thread wchar_t *last_global_error_str = NULL;
```

Comment:
this `__thread` actually bugs me very much:
we claim that it is [thread safe](https://github.com/libusb/hidapi/blob/8741697359b8dd0d73702a69c54e64ffc552d3cf/hidapi/hidapi.h#L461), but imagine next scenario:

thread1: `hid_init()` - success
thread1: `hid_enumerat()` - error happened and registered
thread2: `hid_enumerat()` - error happened and registered
thread1: `hid_exit()` - `last_global_error_str` from thread1 is freed, but `last_global_error_str` from thread2 is leaked

I'm sure the very same issue is present with `hidraw` implementation of the global error

---
your intention with this PR is great, but I'd like to hear some ideas of how can we fix the global issues all together

Reply:
I am not a threading expert, and only copied code from the other implementation from this code base.
Why is thread 2 allowed to run `hid_enumerate()` without calling `hid_init()`

How do we add `register_global_error(NULL);` to a `pthread_cleanup_push`?
But it looks like that's the solution.  In thread we need the cleanup_pop to have ability to clear the hid thread error.

It also looks like I might have missed adding the freeing of the device pointers on close of device

Reply:
> Why is thread 2 allowed to run `hid_enumerate()` without calling `hid_init()`

That is the design of the API. `hid_init()` - is a _global_ initialisation of the library.

> How do we add `register_global_error(NULL)`; to a `pthread_cleanup_push`?
But it looks like that's the solution. In thread we need the cleanup_pop to have ability to clear the hid thread error.

`pthread_cleanup_push` might not be enough, as it clears the resources at thread exit only, and we should come up with something at `hid_exit()`.

I think we need to use [`pthread_key`](https://pubs.opengroup.org/onlinepubs/009695399/functions/pthread_key_create.html) here, unless someone has other ideas.

> It also looks like I might have missed adding the freeing of the device pointers on close of device

Self-code-review is always a good thing.

Reply:
After going thru the git history again, and trying to find a good solution,
I think the best approach would be to discard the faulty behavior introduced not so long ago (https://github.com/libusb/hidapi/pull/23) and don't make any thread-safety assumptions for `hid_error`, i.e.: make `last_global_error_str` a simple `static` variable.